packages/core/test/financial-templates/common/financial-product-libraries/contract-for-difference-libraries/CallOptionContractForDifferenceFinancialProductLibrary.js

```diff
@@ -0,0 +1,89 @@
+);
+
+// helper contracts. To test CFD libraries we simply need a financial contract with an `expirationTimestamp` method.
+
+const ExpiringContractMock = artifacts.require("ExpiringMultiPartyMock");
+
+const { toWei, toBN, utf8ToHex } = web3.utils;
+const strikePrice = toWei("400");
+
+contract("CallOptionContractForDifferenceFinancialProductLibrary", function () {
+  let callOptionCFDFPL;
+  let expiringContractMock;
+
+  beforeEach(async () => {
+    callOptionCFDFPL = await CallOptionContractForDifferenceFinancialProductLibrary.new();
+    expiringContractMock = await ExpiringContractMock.new(
+      ZERO_ADDRESS, // _financialProductLibraryAddress
+      "1000000", // _expirationTimestamp
+      { rawValue: toWei("1.5") }, // _collateralRequirement
+      utf8ToHex("TEST_IDENTIFIER"), // _priceIdentifier
+      ZERO_ADDRESS // _timerAddress
+    );
+  });
+  describe("Contract For diffrence Paramaterization", () => {
+    it("Can set and fetch valid strikes", async () => {
+      await callOptionCFDFPL.setContractForDifferenceParameters(expiringContractMock.address, strikePrice);
+
+      const setStrike = await callOptionCFDFPL.contractForDifferenceStrikePrices(expiringContractMock.address);
+      assert.equal(setStrike.toString(), strikePrice);
+    });
+    it("Can not re-use existing CFD contract address", async () => {
+      await callOptionCFDFPL.setContractForDifferenceParameters(expiringContractMock.address, strikePrice);
+
+      // Second attempt should revert.
+      assert(
+        await didContractThrow(
+          callOptionCFDFPL.setContractForDifferenceParameters(expiringContractMock.address, strikePrice)
+        )
+      );
+    });
+    it("Can not set invalid CFD contract address", async () => {
+      // CFD Address must implement the `expirationTimestamp method.
+      assert(await didContractThrow(callOptionCFDFPL.setContractForDifferenceParameters(ZERO_ADDRESS, strikePrice)));
+    });
+  });
+  describe("Compute expirary tokens for collateral", () => {
+    beforeEach(async () => {
+      await callOptionCFDFPL.setContractForDifferenceParameters(expiringContractMock.address, strikePrice);
+    });
+    it("Lower than strike should return 0", async () => {
+      const expiraryTokensForCollateral = await callOptionCFDFPL.computeExpiraryTokensForCollateral.call(toWei("300"), {
+        from: expiringContractMock.address,
+      });
+      assert.equal(expiraryTokensForCollateral.toString(), toWei("0"));
+    });
+    it("Higher than strike correct value", async () => {
+      const expiraryTokensForCollateral = await callOptionCFDFPL.computeExpiraryTokensForCollateral.call(toWei("500"), {
+        from: expiringContractMock.address,
+      });
+      assert.equal(expiraryTokensForCollateral.toString(), toWei("0.2"));
+    });
+    it("Arbitary expirary price above strike should return correctly", async () => {
+      for (const price of [toWei("500"), toWei("600"), toWei("1000"), toWei("1500"), toWei("2000")]) {
```

Comment:
tiny nit: rather than replicating the computation off-chain, it might be preferable to have this be an array of objects with trivially verifiable output values:
`[{ input: toWei("500"), expectedOutput: toWei("1234") },...]`. If it's hard to create trivially verifiable values, keeping it the way it is could make sense. Totally up to you -- just wanted to mention it.